Add capitalize_first method

glaucocustodio opened the pull request:
Ruby's  `capitalize` method turns all characteres in lowercase, except the first one, which turns uppercase. The `capitalize_first` method turns the first character uppercase and keeps the others as it is.

`"what a Lovely Day".capitalize # => "What a lovely day"`

`"what a Lovely Day".capitalize_first # => "What a Lovely Day"`

If another method is not good, we can add an option to `capitalize` maybe.

jcoleman committed:
Fix button_to's params option to support nested names.

In e6e0579defcfcf94ef1c4c1c7659f374a5335cdb the `params` option was added to the `button_to` helper. However, the patch doesn't support nested hashes so `{a: {b: 'c'}}` for example gets turned into a hidden form input with the name 'a' and the value being the string representation of the `{b: 'c'}` nested hash.

Since Rails supports nested hashes everywhere else (and even in the URL params of link_to and button_to), I believe this to be a bug/unfinished feature.

tomprats committed:
Update session to have indifferent access

st0012 committed:
Porting ActionController::Caching to ActionMailer::Caching

st0012 committed:
Move caching/fragments in ActionMailer and ActionController to action_dispatch/caching/fragments

st0012 committed:
Move most caching methods to ActionDispatch::Caching, and let ActionMailer and ActionController to include it

st0012 committed:
Make caching configuration more flexible

st0012 committed:
Move ActionMailer::Caching's content into ActionMailer::Base instead of including it

Remove useless helper in ActionDispatch::Caching and fix indentation

st0012 committed:
Remove unnecessarily included modules in ActionController::Caching

st0012 committed:
Change ActionMailer's default caching configuration and update generator's environment templates

tomprats committed:
Update Session to utilize indiffernt access

prathamesh-sonpatki committed:
Correctly show deprecation warning for incompatible migrations

maclover7 committed:
Enable Action Cable routes by default

This also marks Action Cable routes as internal to Rails.

kamipo committed:
Fix `assert_in_delta` test failure

`assert_in_delta` in `timestamp_test.rb` causes an intermittent test
failure. It looks like that caused by subseconds truncation in MySQL 5.5.

Example:

```
  1) Failure:
  TimestampTest#test_touching_many_attributes_updates_them [/home/travis/build/rails/rails/activerecord/test/cases/timestamp_test.rb:125]:
  Expected |2016-02-06 09:22:10 +0000 - 2016-02-06 09:22:09 UTC| (1.000837185) to be <= 1.
```

maclover7 committed:
Generate `config/cable.yml` in `rails:update` command

maclover7 committed:
Generate ApplicationRecord if it does not already exist

maclover7 committed:
Generate ApplicationCable files if they do not already exist

maclover7 committed:
Add notes for future selves

Once RubyGems 2.5.0 is required, then the duplicated files can be
removed, and symlinks can be used instead.

[ci skip]

piotrj committed:
Fix logger silencing for broadcasted loggers

Fix #23609

Commit 629efb6 introduced thread safety to logger silencing but it
didn't take into account the fact that the logger can be extended with
broadcasting to other logger.

This commit introduces local_level to broadcasting Module which enables
broadcasted loggers to be properly silenced.

maclover7 committed:
Fix `request.ssl?` bug with Action Cable

This bug affects `wss://` requests when running Action Cable in-app.
Fixes #23620.

maclover7 committed:
Fix issue #23625

This resolves a bug where if the primary key used is not `id` (ex:
`uuid`), and has a `validates_uniqueness_of` in the model, a uniqueness error
would be raised. This is a partial revert of commit `119b9181ece399c67213543fb5227b82688b536f`, which introduced this behavior.

prathamesh-sonpatki committed:
Fix random failures of tests on Travis

- Tests on Travis are randomly failing because schema_migrations table
  does not exist in teardown block.
- Also checked that all other places where we have used
  `ActiveRecord::SchemaMigration.delete_all` we have rescued it, so used
  it here also. This failure was not specifically related to the test
  added in this PR but to overall compatibility migration tests, so
  adding as separate commit.

Commit:
Fix NoMethodError preparable for Arel::Visitors in case prepared statements is falsy

tenderlove committed:
introduce a `tree` factory method for creating the dep tree

also add an EMPTY sentinel node

tenderlove committed:
pull template check up to match existing behavior

also remove the EMPTY node since we won't need it

maclover7 committed:
Full Action Cable documentation read through

This PR checks all active Action Cable documentation for typos and other
fixes. It aims to make sure that when Rails 5 is released, that the
Action Cable docs are up to snuff with the other documentation included
with Rails.

[ci skip]

maclover7 committed:
Add `internal` attribute to routes

This is meant to provide a way for Action Cable, Sprockets, and possibly
other Rack applications to mark themselves as internal, and to exclude
themselves from the routing inspector, and thus `rails routes` / `rake
routes`.

I think this is the only way to have mounted Rack apps be marked as
internal, within AD/Journey. Another option would be to create an array
of regexes for internal apps, and then to iterate over that everytime a
request comes through. Also, I only had the first `add_route` method set
`internal`'s default to false, to avoid littering it all over the
codebase.

ShikChen committed:
Improve the performance of string xor operation

Use `each_byte` instead of `bytes` to speed up string xor operation and
reduce object allocations.

Inspired by commit 02c3867882d6d23b10df262a6db5f937ca69fb53.

``` ruby
require 'benchmark/ips'
require 'allocation_tracer'

a = 32.times.map { rand(256) }.pack('C*')
b = 32.times.map { rand(256) }.pack('C*')

def xor_byte_strings1(s1, s2)
  s1.bytes.zip(s2.bytes).map { |(c1,c2)| c1 ^ c2 }.pack('c*')
end

def xor_byte_strings2(s1, s2)
  s2_bytes = s2.bytes
  s1.bytes.map.with_index { |c1, i| c1 ^ s2_bytes[i] }.pack('c*')
end

def xor_byte_strings3(s1, s2)
  s2_bytes = s2.bytes
  s1.each_byte.with_index { |c1, i| s2_bytes[i] ^= c1 }
  s2_bytes.pack('C*')
end

fail if xor_byte_strings1(a, b) != xor_byte_strings2(a, b)
fail if xor_byte_strings1(a, b) != xor_byte_strings3(a, b)

Benchmark.ips do |x|
  x.report('xor_byte_strings1') { xor_byte_strings1(a, b) }
  x.report('xor_byte_strings2') { xor_byte_strings2(a, b) }
  x.report('xor_byte_strings3') { xor_byte_strings3(a, b) }
  x.compare!
end

Tracer = ObjectSpace::AllocationTracer
Tracer.setup(%i{type})
p xor_byte_strings1: Tracer.trace { xor_byte_strings1(a, b) }
p xor_byte_strings2: Tracer.trace { xor_byte_strings2(a, b) }
p xor_byte_strings3: Tracer.trace { xor_byte_strings3(a, b) }
```

```
Warming up --------------------------------------
   xor_byte_strings1    10.668k i/100ms
   xor_byte_strings2    11.814k i/100ms
   xor_byte_strings3    13.139k i/100ms
Calculating -------------------------------------
   xor_byte_strings1    116.667k (± 3.1%) i/s -    586.740k
   xor_byte_strings2    129.932k (± 4.3%) i/s -    649.770k
   xor_byte_strings3    142.506k (± 4.2%) i/s -    722.645k

Comparison:
   xor_byte_strings3:   142506.3 i/s
   xor_byte_strings2:   129932.4 i/s - 1.10x slower
   xor_byte_strings1:   116666.8 i/s - 1.22x slower

{:xor_byte_strings1=>{[:T_ARRAY]=>[38, 0, 0, 0, 0, 0], [:T_STRING]=>[2, 0, 0, 0, 0, 0]}}
{:xor_byte_strings2=>{[:T_ARRAY]=>[3, 0, 0, 0, 0, 0], [:T_DATA]=>[1, 0, 0, 0, 0, 0], [:T_IMEMO]=>[2, 0, 0, 0, 0, 0], [:T_STRING]=>[2, 0, 0, 0, 0, 0]}}
{:xor_byte_strings3=>{[:T_ARRAY]=>[1, 0, 0, 0, 0, 0], [:T_DATA]=>[1, 0, 0, 0, 0, 0], [:T_IMEMO]=>[2, 0, 0, 0, 0, 0], [:T_STRING]=>[2, 0, 0, 0, 0, 0]}}
```

kaspth committed:
Make collection caching explicit.

Having collection caching that wraps templates and automatically tries
to infer if they are cachable proved to be too much of a hassle.

We'd rather have it be something you explicitly turn on.

This removes much of the code and docs to explain the previous automatic
behavior.

This change also removes scoped cache keys and passing cache_options.

tenderlove committed:
fix recursive templates

tenderlove committed:
clean up classes a little

tenderlove committed:
convert structs to a regular class

tenderlove committed:
change internal implementation to use a tree

tenderlove committed:
remove dead code

this commit removes unused code and changes the monitor to a mutex.
Since the digest doesn't recurse on itself anymore, we can just use a
mutex

tenderlove committed:
remove more dead code

tenderlove committed:
fix rake tasks

bquorning committed:
Tests for AC::Parameters#==

bquorning committed:
Fix AC::Parameters#== with other AC::Parameters

Creating a protected getter method for `@parameters`.

bquorning committed:
Deprecate AC::Parameters#== with a Hash

tenderlove committed:
convert `digest` to only use kwargs

schneems committed:
Alternative to #23638 log to STDOUT via env var

People who deploy to containers or other places where they might have some sort of a log wrapping service use stdout. With this change new rails apps can be configured to output to STDOUT via setting `RAILS_LOG_TO_STDOUT` to any value. This allows container images or services to set the value for all apps without having to modify configuration for each application. If an app wants to opt out, they can either delete from the env hash, or remove that configuration. 

cc/ @rafaelfranca

javan committed:
Add client-side console logging to help debug reconnect issues

maclover7 committed:
Remove unused parameter from method

The `attribute` parameter is not used inside the `normalize_detail`
method. This does not need to go through a deprecation cycle, since the
method is private.

chezou committed:
Add assertions order by field with empty data

Add assertions to MySQL `ORDER BY FIELD()` with empty data.
These tests examine to sanitize `ORDER BY FIELD()` with empty data
appropriately.

```ruby
  Tag.order(['field(id, ?)', []]).to_sql
  # => SELECT "tags".* FROM "tags" ORDER BY field(id, NULL)

  Tag.order(['field(id, ?)', nil]).to_sql
  # => SELECT "tags".* FROM "tags" ORDER BY field(id, NULL)
```

vipulnsward committed:
Fixed passing of delete method on button_to tag, creating wrong form csrf token

Fixes #23524

tenderlove committed:
Merge branch 'master' into treewip

* master: (27 commits)
  move digest cache on to the DetailsKey object
  remove object `hash` cache
  [ci skip] fix suggested change-replace 'an' with 'the in Rails engine guide'
  Missing documentation about hash algorithm option for MessageVerifier [ci skip]
  set `skip_listen` option to dummy appplication
  Fix the language in engines guide
  Add accidentally removed `#` [ci skip]
  fields_for_style needs to test for AC::Parameters
  Fix indentation for code block in changelog
  Remove accidentally duplicated change log title [ci skip]
  partially revert 69009f4473637a44ade26d954ef5ddea6ff903f2
  Remove needless `case_insensitive_comparison` in mysql2 adapter
  modify to `error` also abort when specify fail fast option
  Implement ActionController::Parameters#inspect
  remove unused method
  Remove unused Journey code
  Add Action Cable CHANGELOG in release notes [ci skip]
  Show proper error message when a non-relation object is passed to AR::Relation#or
  Fix semantics of test names for finish option in batches_test
  Fix typo
  ...

Conflicts:
	actionview/lib/action_view/digestor.rb

tenderlove committed:
cache all subnodes in the digest cache

this is for backwards compatibility.  We should remove this after
5-0-stable is created because digest calculation should be amortized.
Caching digests of subtrees will speed up digests for other trees that
share the same children.  However, this will also increase memory, and
after the app warms up, then those subtrees will never be reused which
means that memory is wasted.

lifo committed:
Treat 'closing' state as closed.

We are seeing cases where the websockets get stuck in the 'closing' state
after a tab has been in background for a while. So lets treat those websockets
as closed.

kaspth committed:
Instrument cached collection renders.

Augments the collection caching with some instrumentation that's logged.

For collections that have been cached like:

```ruby
<%= render partial: 'notifications/notification', collection: @notifications, cached: true %>
```

We'll output a line showing how many cache hits we had when rendering it:

```
  Rendered collection of notifications/_notification.html.erb [0 / 100 cache hits] (3396.5ms)
```

tenderlove committed:
use the class level digest method for calculating the digest

tenderlove committed:
push injected dependencies up to the `digest` method

Only the root node has injected dependencies, so we don't need to care
about them at lower levels.  This change pushes the injected
dependencies up to where the user passed them in.

tenderlove committed:
push partial name regexp up

we know that all child templates will be considered as "partials", so
the only one that needs detection by name is the root node

kamipo committed:
Remove `alias exec_without_stmt exec_query`

This alias was for compatibility with legacy mysql adapter.
But the return value of both methods is already inconsistent.

`exec_query` returns `ActiveRecord::Result` instance.
But `exec_without_stmt` returns `[result_set, affected_rows]`

https://github.com/rails/rails/blob/v4.2.5.1/activerecord/lib/active_record/connection_adapters/mysql_adapter.rb#L335-L364

Legacy mysql adapter was already removed in Rails 5.0.
I think we can remove this inconsistent alias.

lifo committed:
Confirm connection monitor subscription on open

lifo committed:
Log ConnectionMonitor connect

akshaymohite committed:
[ci skip] Need to mention debug_exception_response_format in the api_app documentation.

- #23771 removed the reference to debug_exception_response_format from the api_app documentation.
- We need to let users know, they have ability to configure debug_exception_response_format in their development environment.
- Added documentation for the same in api_app.md file
- Grammar corrections

Commit:
Adding test to verify the last week when the year is leap

Fixing failing specification for verifying the last week when the year is leap

javan committed:
Connection#isAlive

javan committed:
Uninstall event handlers when replacing WebSocket instance

Ensures we don't get "onclose" events from a previous WebSocket that was in the "closing" state

javan committed:
ActionCable.log

javan committed:
Fix isAlive condition and add more logging

glaucocustodio committed:
Add capitalize_first method

glaucocustodio committed:
Add capitalize_first method

tadeusrox commented on activesupport/lib/active_support/core_ext/string/inflections.rb:
```diff
@@ -222,6 +222,13 @@ def humanize(options = {})
 
+  # Converts just the first character to uppercase.
+  #
+  #  'what a Lovely Day'.capitalize # => "What a Lovely Day"
```
`'what a Lovely Day'.capitalize` => `'what a Lovely Day'.capitalize_first` ?

glaucocustodio replied:
Thank you, I've fixed.

glaucocustodio committed:
Fix method name in documentation

glaucocustodio committed:
Fix method name in documentation

glaucocustodio committed:
Fix method name in documentation

kdough committed:
Add methods to array delegation from `Relation`

Delegation of some `Array` methods was removed in commit 9d79334. That
change did add explicit delegation of a few methods that `Array` has but
which aren't on `Enumerable`.  However, a few non-mutation methods were
omitted.  This adds `Array` delegation of `#in_groups`, `#in_groups_of`,
`#shuffle` and `#split`.  This allows things like
`MyThing.all.in_groups_of(3) { ... }` to continue working as they did
before commit 9d79334.

rafaelfranca committed:
Merge pull request #17043 from jcoleman/fix-nested-params-in-button-to

Fix button_to's params option to support nested names.

rafaelfranca committed:
Merge pull request #23782 from kdough/add-array-delegation-methods-to-relation

Add methods back to array delegation from ActiveRecord::Relation

rafaelfranca committed:
Merge pull request #23733 from bquorning/actioncontroller-parameters-equality

Deprecate AC::Parameters#== with Hash

egilburg commented on activesupport/lib/active_support/inflector/methods.rb:
```diff
@@ -140,6 +140,10 @@ def humanize(lower_case_and_underscored_word, options = {})
       result
     end
 
+    def capitalize_first string
```
convention is to use parentheses in method names, `def capitalize_first(string)`

glaucocustodio replied:
Thanks.

rafaelfranca commented:
@fxn what do you think?

egilburg commented:
:+1: on the feature, we had this requirement in our codebase too, implemented as:

```ruby
  self[0].upcase + self[1..-1]
```

Here's a benchmark:

```ruby
class String
  def upcase_range
    self[0].upcase + self[1..-1]
  end

  def upcase_sub
    sub(/\S/, &:upcase)
  end
end

require 'benchmark'

n = 1_000_000
str = 'what a Lovely Day'

Benchmark.bmbm do |x|
  x.report('#upcase_range') do
    n.times { str.upcase_range }
  end

  x.report('#upcase_sub') do
    n.times { str.upcase_sub }
  end
end
```

Result:

```
Rehearsal -------------------------------------------------
#upcase_range   0.390000   0.000000   0.390000 (  0.393900)
#upcase_sub    1.380000   0.000000   1.380000 (  1.380069)
---------------------------------------- total: 1.770000sec

                    user     system      total        real
#upcase_range   0.390000   0.000000   0.390000 (  0.388639)
#upcase_sub    1.420000   0.010000   1.430000 (  1.419239)
```

Seems the index+range implementation is about 3.5 times faster than with sub.

eval committed:
Show correct return types for examples [ci skip]

tenderlove committed:
eliminate warnings about multiple primary keys on habtm join tables

habtm join tables commonly have two id columns and it's OK to make those
two id columns a primary key.  This commit eliminates the warnings for
join tables that have this setup.

  ManageIQ/manageiq#6713

tenderlove committed:
Merge branch 'master' into treewip

* master:
  eliminate warnings about multiple primary keys on habtm join tables
  Add methods to array delegation from `Relation`
  Fix button_to's params option to support nested names.
  Dependencies clean up
  Deprecate AC::Parameters#== with a Hash
  Fix AC::Parameters#== with other AC::Parameters
  Tests for AC::Parameters#==
  better docs for ActiveSupport::TestCase#assert_nothing_raised
  remove needless `debug_exception_response_format` config [ci skip]
  partial pass over the API guide [ci skip]
  Fix `unsubscribed` server side behavior
  Use a semaphore to signal message availability
  Fix master build
  Remove github gems from the master bug report templates
  Truncate ActionCable broadcast message to 300 chars
  Remove unused require

akshaymohite commented:
Here is `benchmark-ips` report showing memory footprint as well.
```
Calculating -------------------------------------
 #upcase_first_index   101.810k i/100ms
   #upcase_first_sub    54.490k i/100ms
-------------------------------------------------
 #upcase_first_index      2.272M (± 8.0%) i/s -     11.301M
   #upcase_first_sub    779.316k (± 7.1%) i/s -      3.923M
```

glaucocustodio commented:
Given that string:  `" what a Lovely Day"`

How would we deal with it? Shall we turn `w` uppercase? `sub(/\S/, &:upcase)` does it, ` self[0].upcase + self[1..-1]` does not.

glaucocustodio committed:
Add parentheses in method name

glaucocustodio committed:
Add parentheses in method name

glaucocustodio committed:
Add parentheses in method name

egilburg commented:
@akshaymohite could you do a footprint on variation of `self[0].upcase.concat(self[1..-1])` ?

egilburg commented:
@glaucocustodio is that really the role of capitalize_first to deal with blank characters? If string has blank spaces IMO it should first be squished.

glaucocustodio commented:
I think the same as you @egilburg, that was only to point out a performance/utility tradeoff.

egilburg commented:
Some more implementations :)

```ruby
class String
  def upcase_first_index_interpolate
    "#{self[0].upcase}"#{self[1..-1])}"
  end

  def upcase_first_index_concat
    self[0].upcase.concat(self[1..-1])
  end

  def upcase_first_index_plus
    self[0].upcase + self[1..-1]
  end

  def upcase_first_dup_replace
    dup = self.dup
    dup[0] = dup[0].upcase
    dup
  end

  def upcase_first_sub
    sub(/\S/, &:upcase)
  end

  def upcase_first_dup_sub
    dup.sub!(/\S/, &:upcase)
  end
end

require 'benchmark'
require 'benchmark/ips'

str = 'what a Lovely Day'.freeze

Benchmark.ips do |x|
  x.config(:time => 5, :warmup => 1)

  x.report('#upcase_first_index_interpolate') do
    str.upcase_first_index_interpolate
  end

  x.report('#upcase_first_index_concat') do
    str.upcase_first_index_concat
  end

  x.report('#upcase_first_index_plus') do
    str.upcase_first_index_plus
  end

  x.report('#upcase_first_dup_replace') do
    str.upcase_first_dup_replace
  end

  x.report('#upcase_first_sub') do
    str.upcase_first_sub
  end

  x.report('#upcase_first_dup_sub') do
    str.upcase_first_dup_sub
  end
end
```

Results:

```
Calculating -------------------------------------
#upcase_first_index_interpolate
                          2.846M (± 1.2%) i/s -     14.250M
#upcase_first_index_concat
                          2.121M (± 2.6%) i/s -     10.641M
#upcase_first_index_plus
                          2.117M (± 2.1%) i/s -     10.615M
#upcase_first_dup_replace
                          1.547M (± 1.9%) i/s -      7.776M
   #upcase_first_sub    690.771k (± 1.8%) i/s -      3.491M
#upcase_first_dup_sub
                        542.632k (± 1.4%) i/s -      2.735M
```

Seems a pretty consistent time-space tradeoff. Given it's linearity, I'd argue most would prioritize time over memory usage, as memory can be increased (and is garbage collected after use) while time is paid by the user or a more complex parallelization algorithm.

Commit:
remove args from assert_nothing_raised in tests

Commit:
update assert_nothing_raised in testing.md

fxn commented:
Which use cases does this method have?

gautamsawhney committed:
Fixed a typo

Commit:
add deprecation warning to assert_nothing_raised and changelog entry

egilburg commented:
@fxn Similar as `capitalize` but without destroying existing capital case of proper nouns. Used for list items where it would be beginning of sentence and thus should start capitalized.

```ruby
"user Joe".capitalize # => "User joe"
"user Joe".capitalize_first # => "User Joe"
```

egilburg commented on activesupport/test/inflector_test.rb:
```diff
@@ -157,6 +157,10 @@ def test_acronyms
   end
 
+  def test_capitalize_first
+    assert_equal "What a Lovely Day", ActiveSupport::Inflector.capitalize_first("what a Lovely Day")
```
please also add test for the delegation, `assert_equal "What a Lovely Day", "what a Lovely Day".capitalize_first`, as that will likely be more common usage than manually using the inflector.

prathamesh-sonpatki committed:
Merge pull request #23788 from gautamsawhney/master

Fixed a typo

prathamesh-sonpatki committed:
Improve CHANGELOG for https://github.com/rails/rails/pull/17043 [ci skip]

meinac committed:
Use symbol of mime type instead of object to get correct parser

After registering new `:json` mime type `parsers.fetch` can't find the mime type because new mime type is not equal to old one. Using symbol of the mime type as key on parsers hash solves the problem.

Closes #23766

qazbnm456 committed:
fix typo

y-yagi committed:
use method provided by minitest

The process of converting `Test` to `String` is already defined in minitest.
I think it is better to use that for the consistency of output content.

ref: https://github.com/seattlerb/minitest/blob/master/lib/minitest/test.rb#L261..L267

eileencodes committed:
Always validate record if validating a virtual attribute

Fixes #23645

When you're using an `attr_accessor` for a record instead of an
attribute in the database there's no way for the record to know if it
has `changed?` unless you tell it `attribute_will_change!("attribute")`.

The change made in 27aa4dd updated validations to check if a record was
`changed?` or `marked_for_destruction?` or not `persisted?`. It did not
take into account virtual attributes that do not affect the model's
dirty status.

The only way to fix this is to always validate the record if the
attribute does not belong to the set of attributes the record expects
(in `record.attributes`) because virtual attributes will not be in that
hash.

I think we should consider deprecating this particular behavior in the
future and requiring that the user mark the record dirty by noting that
the virtual attribute will change. Unfortunately this isn't easy because
we have no way of knowing that you did the "right thing" in your
application by marking it dirty and will get the deprecation warning
even if you are doing the correct thing.

For now this restores expected behavior when using a virtual attribute
by always validating the record, as well as adds tests for this case.

I was going to add the `!record.attributes.include?(attribute)` to the
`should_validate?` method but `uniqueness` cannot validate a virtual
attribute with nothing to hold on to the attribute. Because of this
`should_validate?` was about to become a very messy method so I decided
to split them up so we can handle it specifically for each case.

eileencodes committed:
Merge pull request #23793 from qazbnm456/master

fix typo

kaspth committed:
Revert "Instrument read_multi".

Reevaluating the log output generated from this instrumentation,
we've found that it wasn't all that useful in practice.

```
Caches multi read:
- views/david/2/4184ab71db6849621a4d8820fcd2c0ad
- views/david/2/4184ab71db6849621a4d8820fcd2c0ad
- views/david/3/4184ab71db6849621a4d8820fcd2c0ad
- views/david/3/4184ab71db6849621a4d8820fcd2c0ad
```

If rendering many templates the output is inscrutable, and it's impossible
to see how many cache misses there were.

Revert ca6aba7f30 and implement a better way later.

matthewd committed:
Mutating the result of Relation#to_a should not affect the relation

Clarifying this separation and enforcing relation immutability is the
culmination of the previous efforts to remove the mutator method
delegations.

Commit:
Make benchmark('something', silence: true) actually work

claudiob committed:
AppGenerator: Replace 'rake' with 'rails_command'

Since Rails 5.0 is switching the Rails command line from 'rake …'
to 'rails …', it makes sense to also replace the `rake` method in
the Rails templates API.

Based on feedback from @matthewd and @kaspth, I chose to replace
`rake` with `rails_command`, which is less confusing than
the alternatives `rails` or `command` or `rails_run` and is not
Thor-reserved word like `task`.

claudiob committed:
Invoke 'rails' instead of 'rake' in AppGenerator

Follows the convention from Rails 5 of invoking tasks on the command-line
with `rails …` rather than `rake …`.

kaspth committed:
Merge pull request #23695 from kaspth/remove-automatic-collection-caching

Make collection caching explicit.

dgynn committed:
Prevent ActionMailer initializer from triggering load of ActionMailer

the after_initialize block has been changed to use the configuration from
`config.action_mailer` rather than `ActionMailer::Base` so that action mailer
is not loaded before it is necessary.
the mailer preview routes setup have been moved out of an `on_load(:action_mailer)`
block.

matthewd committed:
Merge pull request #23794 from matthewd/mutation-safe-to_a

Mutating the result of Relation#to_a should not affect the relation

rafaelfranca committed:
Merge pull request #23791 from prathamesh-sonpatki/fix-changelog

Improve CHANGELOG for https://github.com/rails/rails/pull/17043 [ci skip]

chashmeetsingh committed:
Fixed typo

prathamesh-sonpatki committed:
Merge pull request #23801 from chashmeetsingh/patch-2

Fixed typo

matthewd commented:
> If another method is not good, we can add an option to capitalize maybe.

I suspect this would be the better choice.

It seems arbitrary that `capitalize_first` means "make the first letter a capital and leave the rest", and `capitalize` means "make the first letter a capital and lowercase the rest": if anything, I think I'd expect them to be the other way around.

A kwarg option allows us to describe the actual difference in behaviour.

fxn commented:
I would like to understand use cases, but the name is something that does not click to me in either case.

The definition of `capitalize` has two parts: what happens to the first character (upcased), and what happens to the rest (lowercased). It could be argued that by saying "first" the name implies "rest" is an empty string and thus logically anything applies to an empty string. But... doesn't click.

The name that would really make sense to me would be `upcase_first`, just upcase the first character (vs upcase all).

kamipo committed:
Reduce `attribute.to_s`

kaspth committed:
Merge pull request #23799 from y-yagi/use_method_provided_by_minitest

use method provided by minitest

kaspth committed:
[ci skip] Mention testing fore-runners.

Get it? ( ͡° ͜ʖ ͡°)

kaspth committed:
[ci skip] Clarify collection caching went EXPLICIT.

Yo dawg, we so explicit if we were a music track, yo' iPhone's Music app would put
an E next to it.

*drops mic*

kaspth committed:
[ci skip] Document collection caching better.

Why are we here? Where did we come from? Where are we going... in 5 minutes?

These questions are not answered in this documentation. Instead collection
caching is better clarified. How it works, how it can share caches with
individual templates.

vipulnsward committed:
Refactored Request Forgery CSRF PerFormTokensController tests and DRY'ed them up.

vipulnsward committed:
Merge pull request #23785 from eval/correct-types-examples

Show correct return types for examples

ianfleeton committed:
Match test name to test (typo)

sgringwe committed:
Add a note about downcasing submit tag

This is a notable change since this will cause confusing test failures for tests relying on the old naming scheme.

spastorino committed:
Merge pull request #23806 from sgringwe/patch-1

Add a note about downcasing submit tag

spastorino committed:
Merge pull request #23805 from ianfleeton/validates-length-of-virtual

Match test name to test (typo)

spastorino committed:
Merge pull request #23743 from maclover7/rm-unused-parameter

Remove unused parameter from method

spastorino committed:
Merge pull request #23803 from kamipo/reduce_attribute_to_s

Reduce `attribute.to_s`

spastorino committed:
Merge pull request #23776 from chi6rag/master

Adding test to verify the last week when the year is leap

spastorino committed:
Merge pull request #23682 from ShikChen/fast_strxor

Improve the performance of string xor operation

spastorino committed:
Merge pull request #23769 from kamipo/remove_alias_exec_without_stmt_exec_query

Remove `alias exec_without_stmt exec_query`

Commit:
Fix typographical error

prathamesh-sonpatki committed:
Merge pull request #23808 from gfvcastro/typographical-error

Fix typographical error

pixeltrix committed:
Merge pull request #23751 from chezou/add-test-case-order-by-field

Add test cases about MySQL ORDER BY FIELD()

matthewd committed:
Merge pull request #22748 from Azzurrio/master

Fix NoMethodError preparable for Arel::Visitors::PostgreSQL

rafaelfranca committed:
Merge pull request #23521 from kamipo/fix_assert_in_delta_failure

Fix `assert_in_delta` test failure

rafaelfranca committed:
Merge pull request #23752 from vipulnsward/23524-fix-button_to_delete

Fixed passing of delete method on button_to tag, creating wrong form csrf token

rafaelfranca committed:
Do not pass more options that are needed.

We only need action and method so pass them explicitly instead of
merging the hash with HTML options.

rafaelfranca committed:
Transform the mime object to symbol when registering the parsers

This will keep our current API working without having the users to
change their codebases.

rafaelfranca committed:
Make sure we unregister the mime type before registering

Also make sure we don't change the global state of our test suite.

rafaelfranca committed:
Make per form token work when method is not provided

When `button_to 'Botton', url` form was being used the per form token
was not correct because the method that is was being used to generate it
was an empty string.

rafaelfranca committed:
We are calling `to_s` in the method so we can call downcase now

tenderlove committed:
make sure `rake test` respects TESTOPTS

We should be able to pass options to minitest via TESTOPTS environment
variable

rafaelfranca committed:
Merge pull request #23816 from rails/params_parser_api

Make sure the parameter parsers register API work with overidden mime types.

rafaelfranca committed:
Merge pull request #23600 from maclover7/actioncable-work

Assorted Railties generators work

rafaelfranca committed:
Merge pull request #23668 from maclover7/cable-docs

Full Action Cable documentation read through

rafaelfranca committed:
Merge pull request #23669 from maclover7/internal-routes

Add `internal` attribute to routes

meinac committed:
Action Cable readme fix [ci skip]

rafaelfranca committed:
Merge pull request #23819 from meinac/actioncable_doc_fix

Action Cable readme fix [ci skip]

matthewd committed:
Merge pull request #23419 from prathamesh-sonpatki/fix-showing-of-deprecation-warning-for-legacy-migrations

Correctly show deprecation warning for incompatible migrations

rafaelfranca committed:
Merge pull request #23821 from dgynn/actionmailer_lazy_loading

Prevent ActionMailer initializer from triggering load of ActionMailer

kamipo committed:
Remove needless `drop_table :test_limits`

A `:test_limits` table has not been created.

rafaelfranca committed:
Merge pull request #23825 from kamipo/remove_needless_drop_table_test_limits

Remove needless `drop_table :test_limits`

rafaelfranca committed:
Make sure we don't change the global state in the tests

We are creating the table but not deleting after the test.

Commit:
Merge pull request #23813 from lifo/faye-websocket

Improve Action Cable reconnection reliability

tenderlove committed:
reestablish previous connection after creating all databases

creating all databases mutates the connection pool.  This patch restores
the connection pool to the connection spec established before creating
all databases.  Fixes #23279

rafaelfranca committed:
Merge pull request #23789 from wisetara/wisetara/deprecate-args-ActiveSupport__TestCase#assert_nothing_raised-for-pr

Wisetara/deprecate args active support  test case#assert nothing raised for pr

chancancode committed:
Lock down new `ImplicitRender` behavior for 5.0 RC

1. Conceptually revert #20276

   The feature was implemented for the `responders` gem. In the end,
   they did not need that feature, and have found a better fix (see
   plataformatec/responders#131).

   `ImplicitRender` is the place where Rails specifies our default
   policies for the case where the user did not explicitly tell us
   what to render, essentially describing a set of heuristics. If
   the gem (or the user) knows exactly what they want, they could
   just perform the correct `render` to avoid falling through to
   here, as `responders` did (the user called `respond_with`).

   Reverting the patch allows us to avoid exploding the complexity
   and defining “the fallback for a fallback” policies.

2. `respond_to` and templates are considered exhaustive enumerations

   If the user specified a list of formats/variants in a `respond_to`
   block, anything that is not explicitly included should result
   in an `UnknownFormat` error (which is then caught upstream to
   mean “406 Not Acceptable” by default). This is already how it
   works before this commit.

   Same goes for templates – if the user defined a set of templates
   (usually in the file system), that set is now considered exhaustive,
   which means that “missing” templates are considered `UnknownFormat`
   errors (406).

3. To keep API endpoints simple, the implicit render behavior for
   actions with no templates defined at all (regardless of formats,
   locales, variants, etc) are defaulted to “204 No Content”. This
   is a strictly narrower version of the feature landed in #19036 and
   #19377.

4. To avoid confusion when interacting in the browser, these actions
   will raise an `UnknownFormat` error for “interactive” requests
   instead. (The precise definition of “interactive” requests might
   change – the spirit here is to give helpful messages and avoid
   confusions.)

Closes #20666, #23062, #23077, #23564

[Godfrey Chan, Jon Moss, Kasper Timm Hansen, Mike Clark, Matthew Draper]

egilburg commented:
:+1: for `upcase_first`, that's how our implementation is named as well.

sgrif committed:
Merge pull request #23628 from maclover7/fix-23625

Fix issue #23625

tenderlove committed:
clear digest cache between tests

otherwise we don't get the log messages we're looking for and the tests
fail intermittently.

schneems committed:
Merge pull request #23734 from schneems/schneems/stdout-again

Alternative to #23638 log to STDOUT via env var

eileencodes committed:
Revert changes to validations from PR #18612

In order to fix issue #17621 we added a check to validations that
determined if a record should be validated. Based on the existing tests
and behavior we wanted we determined the best way to do that was by
checking if `!record.peristed? || record.changed? || record.marked_for_destruction?`

This change didn't make it into a release until now. When #23790 was
opened we realized that `valid?` and `invalid?` were broken and did not
work on persisted records because of the `!record.persisted?`.

While there is still a bug that #17621 brought up, this change was too
drastic and should not be a RC blocker. I will work on fixing this so
that we don't break `valid?` but also aren't validating parent records
through child records if that parent record is validate false. This
change removes the code changes to validate and the corresponding tests.
It adds tests for two of the bugs found since Rails 5 beta2 release.

Fixes #17621

tenderlove committed:
remove useless method

rafaelfranca committed:
Move Caching module to Abstract Controller

Abstract Controller is the common component between Action Mailer and
Action Controller so if we need to share the caching component it need
to be there.

sstephenson committed:
Switch to `data-turbolinks-track="reload"`

sstephenson committed:
New apps get the Turbolinks beta gem

sstephenson committed:
Update to the latest Turbolinks gems

sstephenson committed:
Update JavaScript guide for Turbolinks 5

groeneman committed:
[ci skip] Fix typo in Form Helpers Guide

kamipo committed:
Fix `assert_nothing_raised` deprecation warning format

Before:

```
DEPRECATION WARNING: Passing arguments to assert_nothing_raised
        is deprecated and will be removed in Rails 5.1.
```

After:

```
DEPRECATION WARNING: Passing arguments to assert_nothing_raised is deprecated and will be removed in Rails 5.1.
```

groeneman committed:
[ci skip] No hyphen in 'side effects'

matthewd committed:
Merge pull request #23632 from maclover7/fix-23620

Fix `request.ssl?` bug with Action Cable

maclover7 committed:
Merge pull request #23837 from mgroeneman/fix-form-helper-guide-typo

Fix typo in Form Helpers Guide

yui-knk committed:
[ci skip] Improve the readability of documents of nested_attributes

maclover7 committed:
Merge pull request #23840 from yui-knk/improve_the_readability_of_ar_doc

[ci skip] Improve the readability of documents of nested_attributes

kamipo committed:
`drop_table :test_text_limits` as well

Follow up to 3b01785.

rafaelfranca committed:
Merge pull request #23838 from kamipo/fix_assert_nothing_raised_deprecation_warning_format

Fix `assert_nothing_raised` deprecation warning format

rafaelfranca committed:
Merge pull request #23836 from sstephenson/turbolinks-5

Turbolinks 5 compatibility changes

rafaelfranca committed:
Use the trick to get beta releases in the Gemfile

rafaelfranca committed:
Merge pull request #23842 from kamipo/drop_table_test_text_limits_as_well

`drop_table :test_text_limits` as well

rafaelfranca committed:
Move private methods to the private visibility

rafaelfranca committed:
Merge branch 'actionmailer-cache'

This is a rebased version of #22825.

Closes #22825.

rafaelfranca committed:
Add changelog entry for #22825

rafaelfranca committed:
Merge pull request #20851 from tomprats/indifferent-sessions

Give Sessions Indifferent Access

y-yagi committed:
correctly check `ApplicationRecord` is exist in moutable engine

Follow up to 1813b29fc7632959800252f36e4b2e6ed4ac7266

kamipo committed:
Ensure `drop_table` even if tests failure or interrupted

I was encountered remaining `:binary_testings` table by tests failure.
When remaining `:binary_testings` table, never reach `drop_table` due to
`create_table` in the test always fails.

prathamesh-sonpatki committed:
Show permitted flag in the output of AC::Parameters#inspect

- Fixes #23822.

prathamesh-sonpatki committed:
Add CHANGELOG for https://github.com/rails/rails/pull/23734 [ci skip]

Commit:
Merge pull request #23795 from claudiob/replace-rack-with-rails-command

AppGenerator: Replace 'rake' with 'rails_command'

aaronang committed:
Fix title of README according to Markdown conventions

The first heading in some README's are indicated using a second level
heading (`##`), which in my opinion is of incorrect structure.
Therefore, in this patch I changed the first heading to a first level
heading (`#`) in README's where this incorrect structure occurs.

[ci skip]

rafaelfranca committed:
Ignore callstacks from Ruby stdlib in deprecation

Fixes #22982.

rafaelfranca committed:
Merge pull request #23849 from prathamesh-sonpatki/fix-23822

Show permitted flag in the output of AC::Parameters#inspect

rafaelfranca committed:
Always use the released coffee-rails in new applications

coffee-rails is stable now so we don't need to point to the master
branch.

rafaelfranca committed:
Merge pull request #23616 from piotrj/issue_23609_logger_broadcast_silencing

Fix logger silencing for broadcasted loggers

rafaelfranca committed:
Make internal class as nodoc

Also use the new hash syntax.

rafaelfranca committed:
No need CHANGELOG entry for #23849.

It is not a released feature so we don't need to add changelogs to
changes on it.

[ci skip]

rafaelfranca committed:
Merge pull request #23850 from prathamesh-sonpatki/changelog-for-23734

Add CHANGELOG for https://github.com/rails/rails/pull/23734 [ci skip]

eileencodes committed:
Remove changelog entry for reverted commit

I had to revert changes made for this CHANGELOG entry so this is no
longer valid. The change for this entry was removed in 2c02bc0.

rafaelfranca committed:
:bomb: run the test @rafaelfranca :angry:

rafaelfranca committed:
Merge pull request #23853 from kamipo/ensure_drop_table

Ensure `drop_table` even if tests failure or interrupted

kaspth committed:
[ci skip] Move collection caching changelog entry.

We changed this in beta2, and only editing the original entry means
people can't see that it was significantly changed.

eileencodes committed:
Prep release for Rails 5 beta3

rafaelfranca committed:
Enable tmp_restart plugin for puma

Fixes #23855

st0012 committed:
Update 5.0 release notes

eileencodes committed:
Preparing for 5.0.0.beta3 release

Adds changelog headers for beta3 release

st0012 committed:
Add ActionMailer configuration options

st0012 committed:
Add caching guide in ActionMailer basics

rafaelfranca committed:
Do not define methods in the included block

Instance methods can be defined in the module itself

st0012 committed:
Fix ActionView's cache section reference

claudiob committed:
Add Ruby formatting to CHANGELOG entry

[ci skip]

st0012 committed:
Fix wording and wrong reference

claudiob committed:
Merge pull request #23858 from claudiob/fix-aca-readme

Add Ruby formatting to CHANGELOG entry

rafaelfranca committed:
Merge pull request #23857 from st0012/update-documents

[ci skip] Update guide sources for mailer view caching feature

prathamesh-sonpatki committed:
Don't put config.action_mailer.perform_caching entry twice in development.rb

rafaelfranca committed:
Merge pull request #23859 from prathamesh-sonpatki/dont-put-perform-caching-config-twice

Don't put config.action_mailer.perform_caching entry twice in development.rb

zerothabhishek committed:
changes caching guide to add note on weak etags

rafaelfranca committed:
Merge pull request #23860 from zerothabhishek/weak-etag-guide

Changes caching guide to add note on weak etags

chashmeetsingh committed:
Updated file documentation [ci skip]

prathamesh-sonpatki committed:
Added a test for generating Strong ETag

prathamesh-sonpatki committed:
Merge pull request #23861 from chashmeetsingh/patch-1

Updated file documentation [ci skip]

eileencodes committed:
Ensure suppressor runs before validations

I ran into an issue where validations on a suppressed record were
causing validation errors to be thrown on a record that was never going
to be saved.

There isn't a reason to run the validations on a record that doesn't
matter.

This change moves the suppressor up the chain to be run on the `save` or
`save!` in the validations rather than in persistence. The issue with
running it when we hit persistence is that the validations are run
first, then we hit persistance, and then we hit the suppressor. The
suppressor comes first.

The change to the test was required since I added the
`validates_presence_of` validations. Adding this alone was enough to
demonstrate the issue. I added a new test to demonstrate the new
behavior is explict.

tenderlove committed:
Merge branch 'master' into treewip

* master: (113 commits)
  remove useless method
  Updated file documentation [ci skip]
  changes caching guide to add note on weak etags
  Don't put config.action_mailer.perform_caching entry twice in development.rb
  Fix wording and wrong reference
  Add Ruby formatting to CHANGELOG entry
  Fix ActionView's cache section reference
  Do not define methods in the included block
  Add caching guide in ActionMailer basics
  Add ActionMailer configuration options
  Preparing for 5.0.0.beta3 release
  Update 5.0 release notes
  Enable tmp_restart plugin for puma
  Prep release for Rails 5 beta3
  [ci skip] Move collection caching changelog entry.
  Ensure `drop_table` even if tests failure or interrupted
  :bomb: run the test @rafaelfranca :angry:
  Remove changelog entry for reverted commit
  Add CHANGELOG for https://github.com/rails/rails/pull/23734 [ci skip]
  No need CHANGELOG entry for #23849.
  ...

abhishekjain16 committed:
[ci skip] Replace usage of rake routes with rails routes

maclover7 committed:
Merge pull request #23866 from abhishekjain16/document

[ci skip] Replace usage of rake routes with rails routes

maclover7 committed:
Only open Action Cable connection if data is being sent

maclover7 committed:
Merge pull request #23870 from akshaymohite/api-app-debug-exception-response-format-in-doc

[ci skip] Need to mention debug_exception_response_format in the api_app documentation.

rafaelfranca committed:
Merge pull request #23862 from prathamesh-sonpatki/test-for-strong-etag

Added a test for generating Strong ETag

matthewd committed:
Merge pull request #23488 from maclover7/actioncable-defaults

Enable ActionCable routes by default

maclover7 committed:
Only hijack Rack socket when first needed

Fixes #23471

maclover7 committed:
Merge pull request #23863 from delftswa2016/fix-readme-heading

Fix README heading according to Markdown conventions

matthewd committed:
Merge pull request #23843 from maclover7/fix-23471

Only hijack Rack socket when first needed

st0012 committed:
Update config.force_ssl's documentation

st0012 committed:
Add config.debug_exception_response_format into configuration guide

chancancode committed:
Merge pull request #23827 from rails/new_implicit_render

Lock down new `ImplicitRender` behavior for 5.0 RC

kaspth committed:
Additional review of 6b31761.

* Fixes typos in error message and release notes.
* Removes unused template test file.

glaucocustodio committed:
Change method name to upcase_first. Change methods implementation to a faster one.

glaucocustodio committed:
Change method name to upcase_first. Change methods implementation to a faster one.

glaucocustodio committed:
Change method name to upcase_first. Change methods implementation to a faster one.

glaucocustodio commented:
`upcase_first` seems better to me, simpler and shorter. I have renamed the method and changed its implementation to the faster one.

pixeltrix committed:
Merge pull request #23874 from st0012/update-documents

[ci skip] Update configuration guide

jrafanie committed:
Fix uniqueness validation with an after_create hook.

record.id_was is nil in after_create/after_save, so we should use
id in these cases.

While this logic feels incomplete, the existing update_record uses the same
logic:
https://github.com/rails/rails/blob/2fda4e0874a97a76107ab9e88305169f2c625933/activerecord/lib/active_record/relation.rb#L83

This logic was originally added for a similar problem:
updates not working with after_create hook.

See: 482f8c15b1d699c95bfbc3d836f674a09c0d9031

Followup to #23581
Fixes #23844

claudiob committed:
Doc: update routes guides for Rails 5

[ci skip]

claudiob committed:
Doc: update AC::Parameters guides for Rails 5

[ci skip]

claudiob committed:
Merge pull request #23889 from claudiob/fix-getting-started

Doc: update guides for Rails 5

matthewd committed:
Revert "Update Session to utilize indiffernt access"

This reverts commit 45a75a3fcc96b22954caf69be2df4e302b134d7a.

HWIAs are better than silently deeply-stringified hashes... but that's a
reaction to a shortcoming of one particular session store: we should not
break the basic behaviour of other, more featureful, session stores in
the process.

Fixes #23884

matthewd committed:
Revert "Merge pull request #20851 from tomprats/indifferent-sessions"

This reverts commit 22db455dbe9c26fe6d723cac0758705d9943ea4b, reversing
changes made to 40be61dfda1e04c3f306022a40370862e3a2ce39.

This finishes off what I meant to do in 6216a092ccfe6422f113db906a52fe8ffdafdbe6.

matthewd committed:
Merge pull request #23887 from jrafanie/fix_uniqueness_validation_on_after_create

Fix uniqueness validation with an after_create hook.

rafaelfranca committed:
Merge pull request #23846 from y-yagi/correctly_check_ApplicationRecord_in_mountable_engine

correctly check `ApplicationRecord` is exist in moutable engine

rafaelfranca commented:
r? @fxn 

@glaucocustodio could you rebase it?

glaucocustodio committed:
Merge branch 'add_capitalize_first_method' of github.com:glaucocustodio/rails into add_capitalize_first_method

glaucocustodio committed:
Merge branch 'add_capitalize_first_method' of github.com:glaucocustodio/rails into add_capitalize_first_method
